Getting started with Python client

bgoesswe opened the pull request:
First attempt of a Python client HowTo.
Added a placeholder for QGIS HowTo.
Also restructured the user documentation menu a little bit.

m-mohr commented on .vuepress/config.js:
```diff
@@ -12,7 +12,11 @@ const versions = [
     ],
     userNav: [
-      {text: 'Getting Started', link: 'getting-started.html'},
+      {text: 'Getting Started', items: [
```
We should avoid to break existing links. Could we put each client into separate folders. I think the link could just be qgis/index.html or so.

bgoesswe replied:
Ok I just (dff2c7e) restructured it to qgis/index.html and python/index.html also moving their images folders in the subfolders.

m-mohr replied:
Thanks.

m-mohr commented on .vuepress/config.js:
```diff
@@ -13,6 +13,8 @@ const versions = [
       {text: 'Getting Started', link: 'getting-started.html'},
+      {text: 'Python Client', link: 'python/index.html'},
+      {text: 'QGIS Plugin', link: 'qgis/index.html'},
       {text: 'Glossary', link: 'glossary.html'},
```
The Glossary should be above the clients, I think.

bgoesswe replied:
Fixed in 
a68b748

m-mohr commented on documentation/1.0/python/index.md:
```diff
@@ -0,0 +1,335 @@
+You need to inform yourself at your backend provider of choice, which authentication approach you have to carry out. 
+You can also have a look at the [openEO Hub](https://hub.openeo.org/) to see the available authentication types of the backends.     
+
+### Basic Authentication
```
Basic shouldn't really be used by end-users in the long run. We can keep it here for now, but maybe move OIDC up and then add a short note to Basic that OIDC is the preferred way of authenticating.

bgoesswe replied:
Ok, swapped the order and added a sentence for the preference in 
a68b748 .

m-mohr commented on documentation/1.0/python/index.md:
````diff
@@ -0,0 +1,335 @@
+datacube = connection.load_collection("COPERNICUS/S1_GRD")
+```
+
+This results in a [datacube object](https://openeo.org/documentation/1.0/glossary.html#spatial-data-cubes) with the complete dataset of the "COPERNICUS/S1_GRD" collection.
````
This should be a relative link

bgoesswe replied:
fixed all relative links in 
a68b748

m-mohr replied:
Except for one, see comment below.

m-mohr commented on documentation/1.0/python/index.md:
```diff
@@ -0,0 +1,335 @@
+## Job Management
+
+Assuming that the definition of the datacube object and all related processes is finished, we can now send it to the backend and start the execution. 
+In openEO, an execution of a process graph (here defined in the datacube object) is called a [job](https://openeo.org/documentation/1.0/glossary.html#data-processing-modes). Therefore, we need to create a job at the backend using our datacube.
```
Relative link

m-mohr commented on documentation/1.0/python/index.md:
````diff
@@ -0,0 +1,335 @@
+It is not part of the implemented functions of the Python client, therefore, we need to define it manually. This might change in future versions of the client. 
+
+```python
+from openeo.internal.graph_building import PGNode
````
Why's this under internal? Sounds like it shouldn't be used?

bgoesswe replied:
Yes, it should not be used, but unfortunately there is no complex apply in the python client at the moment so this is how you can work around it, I thought it is maybe also interesting for the user how you can work around that, but I updated that part with a less "internal" approach.
Still, this needs updates when the python client improves.

m-mohr replied:
Yeah, that should be improved. Is there a Python Client issue to track this?

bgoesswe replied:
Yes https://github.com/Open-EO/openeo-python-client/issues/130
But I will anyways need to look into that soon because I need that fixed for our next deliverable...

m-mohr commented on documentation/1.0/python/index.md:
````diff
@@ -0,0 +1,335 @@
+```python
+job = datacube.send_job()
+
+res = job.start_and_wait().download_results("/tmp")
````
For a getting started, there should be a hint that this only works on Linux and not on Windows and it should be described how to work on both systems.

bgoesswe replied:
Ok changed that in 
a68b748

m-mohr replied:
Now it says `job.start_and_wait().download_results()`, but there's no parameter specified. Does it default to the temp directories of the OS? Otherwise the code would be invalid?!

bgoesswe replied:
Yes, it defaults to the temporary folder of the OS, but I also added a sentence that you can also provide a file path.

m-mohr commented on documentation/1.0/python/index.md:
```diff
@@ -0,0 +1,335 @@
+
+![Python client RGB composite](./images/example_result.png "RGB composite")
+
+The [source code](https://github.com/Open-EO/openeo-python-client/blob/master/examples/gee_uc1_temp.py) of this example can be found on GitHub.
```
I think this should link to a 1.0 client tag, because master at some point could be API 2.0 and then the link gets confuding.

bgoesswe replied:
Unfortunately, there is no version 1.0 tag yet, so that needs to be done in the future, for now I changed it to the commit that has this example version. 
a68b748

m-mohr commented on documentation/1.0/python/index.md:
```diff
@@ -0,0 +1,335 @@
+
+## Additional Information
+
+* [Python Client Examples](https://github.com/Open-EO/openeo-python-client/tree/master/examples)
```
"Python Client " is pretty redundant here.

bgoesswe replied:
done in 
a68b748

m-mohr commented on documentation/1.0/qgis/index.md:
```diff
@@ -0,0 +1,253 @@
+If logged in successfully, the main window of the plugin appears with three tabs (**Backend Info**, **Jobs**, **Services**). 
+The following sections will go through the capabilities of these tabs. 
+
+## Tab1: Backend Info
```
I think the "TabX: " parts can be removed from the headers

bgoesswe replied:
done in
a68b748

m-mohr commented on documentation/1.0/qgis/index.md:
```diff
@@ -0,0 +1,253 @@
+To learn more about the process, you can click on the info button next to the drop down menu 
+(marked red in the figure below). The return type of the process is written next to the info box. If you want to get a
+good overview of all processes of openEO you might visit the 
+[official processes page](https://openeo.org/documentation/1.0/processes.html), but note that not all backends 
```
Relative links also in this document.

bgoesswe replied:
done in 
a68b748

m-mohr commented on documentation/1.0/qgis/index.md:
```diff
@@ -0,0 +1,253 @@
+additional meta-information (e.g. title, creation date). The first two columns show the title and the creation date of 
+the job. The other elements of the table are numbered in the screenshot and described in the following. 
+
+![QGIS Jobs](./images/jobs.png "Jobs tab")
```
Wait, is that valid markdown? The "Jobs tab" in here confuses me...

bgoesswe replied:
Yes it is valid actually, this is the text that will show up if you slide with the mouse over the image (see https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#images).

m-mohr replied:
But is it really needed? The text in square brackets is already doing the same thing?!

m-mohr commented on documentation/1.0/qgis/index.md:
```diff
@@ -0,0 +1,253 @@
+It contains information like a description, a cost plan etc., of course only if this information is provided by the backend.
+6. **Delete** - By clicking on the "Trash" button the job gets deleted from the backend, after being asked if you are 
+really sure about it.
+7. **Load Job from openEO Hub** - This button opens a new window with a list of shared jobs (or process graphs) from the openEO Hub 
```
Is that still working? The Hub has disabled the Exchange section and it will be updated to 1.0 soon...

bgoesswe replied:
Yes, it still works. In the Hub API, the process graph exchange is still available. But I will update the plugin with the new 1.0 version.

m-mohr replied:
Although nobody can really relate to it at the moment as the Hub doesn't show anything at the moment. I guess it would be a good idea to remove it for now.

bgoesswe replied:
Ok removed the references in
84e9c23 
In the button description, I just said it is not recommended  to use it since I can not easily remove the button from the official plugin version (would take some days to be approved again)

m-mohr commented on documentation/1.0/qgis/index.md:
```diff
@@ -0,0 +1,253 @@
+the [openEO web editor](https://editor.openeo.org/). There you can create new services via the editor or from your jobs.
+6. **Refresh Table** - This button refreshes the list of services manually.
+
+[//]: <> (## Example Job Adaption To Do: Create Youtube Walk-Through)
```
Here's something missing, right? Use HTML comments to not show such things publicly.

bgoesswe replied:
Removed it, was just a reminder  ^^

m-mohr commented on documentation/1.0/qgis/index.md:
```diff
@@ -0,0 +1,253 @@
+
+## Additional Information
+
+* [openEO QGIS Plugin Github](https://github.com/Open-EO/openeo-qgis-plugin)
```
Could probably be just "GitHub repository" and "QGIS Plugin Page" or so

bgoesswe replied:
done in 
a68b748

m-mohr commented on documentation/1.0/getting-started.md:
```diff
@@ -1,20 +1,20 @@
 
 ## Contribute
 
-Didn't find your programming language? You can also access the [openEO API](./developers/api/reference.md) implementations directly or start [implementing your own client library](./developers/clients/getting-started.md).
```
I'm not exactly sure why, but vuepress recommends to put the ./ in front of the links.

bgoesswe replied:
done in
a68b748

m-mohr commented on documentation/1.0/python/index.md:
````diff
@@ -0,0 +1,339 @@
+import openeo
+```
+If this gives you the following output, something went wrong with the installation, please check again the requirements. 
+If you have still troubles installing the Python module, feel free to [contact us](https://openeo.org/contact.html) or leave an issue at the [GitHub project](https://github.com/Open-EO/openeo-python-client/issues).
````
Still a absolute link

bgoesswe replied:
sry, fixed in 2a541c2

m-mohr commented on documentation/1.0/python/index.md:
```diff
@@ -0,0 +1,339 @@
+
+If you do not know an openEO backend that you want to connect to yet, you can have a look at the [openEO Hub](https://hub.openeo.org/), to find all known backends with information on their capabilities.
+
+For this tutorial we will use the Google Earth Engine openEO driver (https://earthengine.openeo.org/v1.0)
```
Url should only be: https://earthengine.openeo.org

bgoesswe replied:
fixed in 2a541c2

m-mohr commented on documentation/1.0/python/index.md:
````diff
@@ -0,0 +1,339 @@
+
+```python
+import openeo
+connection = openeo.connect("https://earthengine.openeo.org/v1.0")
````
Url should only be https://earthengine.openeo.org

bgoesswe replied:
fixed in 2a541c2

m-mohr commented on documentation/1.0/qgis/index.md:
```diff
@@ -47,7 +47,7 @@ The following sections will go through the capabilities of these tabs.
 ## Exploring a backend
 
 The first tab shows information about the backend and the available collections and processes. 
-On the top, there is a basic description about the backend you are connected to.
```
Hehe, that should not be captialized.

m-mohr commented on documentation/1.0/python/index.md:
````diff
@@ -0,0 +1,339 @@
+
+```python
+# Download job results
+job.download_results("download_path")
````
If people copy paste that it does what? Should it just be `job.download_results()` saying that it stores in the current dir?

m-mohr commented on documentation/1.0/python/index.md:
````diff
@@ -0,0 +1,339 @@
+
+```python
+# Starts the job and waits until it finished to download the result.
+job.start_and_wait().download_results("download_path")
````
If people copy paste that it does what? Should it just be `job.download_results()` saying that it stores in the current dir?

m-mohr commented on documentation/1.0/python/index.md:
````diff
@@ -0,0 +1,339 @@
+```
+
+Now the resulting PNG file of the RGB backscatter composite is in your current directory. 
+You can also specify a different output directory in the `download_results(OUTPUT_DIR)` call.
````
That's a bit "cryptic". Could we improve that to something like  "You can specify a different output directory by passing the path as first parameter in the `download_results` call."?

bgoesswe commented:
Ok I re-arranged the download_result a bit so that the mentioning of the "download path" parameter is in the general explanation and not in the example. I also removed the path everywhere else and mention the current dir. Changed the style of functions in the text, to differ from other semicolons

m-mohr commented:
Thanks, let's see how it looks deployed :-)

m-mohr commented:
@bgoesswe I think I'd like the menu items to just be "Python" and "QGIS". What do you think?

m-mohr commented:
That's not an issue as it would be in the separate Developers menu. I'll do the change then.
